radicle-node/src/worker.rs

```diff
@@ -262,84 +200,96 @@ impl<G: Signer + Ecdh + 'static> Worker<G> {
-            // Enable atomic fetch. Only works with Git 2.31 and later.
-            cmd.arg("--atomic");
-        }
+        let staging = match res.and_then(|()| staging.into_staging_refs().map_err(FetchError::from))
+        {
+            Ok(staging) => staging,
+            Err(e) => return (session, Err(e)),
+        };
 
-        // Ignore our own remote when fetching
-        let mut fetchspecs = namespaces.as_fetchspecs();
-        fetchspecs.push(format!("^refs/namespaces/{}/*", self.local));
+        let (session, res) = self.tunnel_fetch(&staging.repo, staging.refspecs(), session);
 
-        cmd.arg(format!("git://{tunnel_addr}/{}", repo.id.canonical()))
-            .args(&fetchspecs)
-            .stdout(process::Stdio::piped())
-            .stderr(process::Stdio::piped())
-            .stdin(process::Stdio::piped());
+        if let Err(e) = res {
+            return (session, Err(e));
+        }
 
-        log::debug!(target: "worker", "Running command: {:?}", cmd);
+        (session, staging.transfer().map_err(FetchError::from))
+    }
 
-        let mut child = cmd.spawn()?;
-        let stderr = child.stderr.take().unwrap();
+    fn upload_pack(
+        &self,
+        fetch: &Fetch,
+        pktline_r: &mut pktline::Reader<WireReader>,
+        stream_w: &mut WireWriter<G>,
+    ) -> Result<Vec<RefUpdate>, FetchError> {
+        match self._upload_pack(fetch, pktline_r, stream_w) {
+            Ok(()) => {
+                log::debug!(target: "worker", "Upload of {} to {} exited successfully", fetch.rid, fetch.remote);
 
-        thread::Builder::new().name(self.name.clone()).spawn(|| {
-            for line in BufReader::new(stderr).lines().flatten() {
-                log::debug!(target: "worker", "Git: {}", line);
+                Ok(vec![])
+            }
+            Err(err) => {
+                log::error!(target: "worker", "Upload error for {}: {err}", fetch.rid);
+
+                // If we exited without receiving a `done` packet, wait for it here.
+                // It's possible that the daemon exited first, or the remote crashed.
+                log::debug!(target: "worker", "Waiting for `done` packet from remote..");
+                let mut header = [0; pktline::HEADER_LEN];
+
+                // Set the read timeout for the `done` packet to twice the configured
+                // value that is used for the fetching (initiator) side.
+                //
+                // This is because the uploader always waits for the `done` packet;
+                // so in case the fetch is aborted by the uploader, eg. if
+                // it can't connect with the daemon, it will wait long enough for the
+                // fetcher to timeout before timing out itself, and will thus receive
+                // the `done` packet.
+                pktline_r
+                    .stream()
+                    .as_connection()
+                    .set_read_timeout(Some(self.timeout * 2))
+                    .ok();
```

Comment:
What is the '.ok()' for?

Reply:
That's pretty common for channel communications, I believe. If the channel is gone away then reading will throw an error anyway and there's nothing we can do about it anyway -- so this `Err` is silence, so to say.